openedx/core/djangoapps/course_roles/helpers.py

```diff
@@ -0,0 +1,189 @@
+#   roles checks. If the flag is not enabled, only the roles checks will be used.
+# .. toggle_use_cases: temporary
+# .. toggle_creation_date: 2023-10-17
+# .. toggle_target_removal_date: 2023-12-01
+# .. toggle_warning:
+USE_PERMISSION_CHECKS_FLAG = WaffleFlag('course_roles.use_permission_checks', module_name=__name__) or False  # lint-amnesty, pylint: disable=toggle-missing-annotation
+
+
+def use_permission_checks():
+    """
+    Returns ture if permissions checks should be used
+    """
+    return USE_PERMISSION_CHECKS_FLAG.is_enabled()
+
+
+@request_cached()
+def user_has_permission_course(
+        user: Union[User, AnonymousUser],
+        permission: Union[CourseRolesPermission, str],
+        course_key: CourseKey
+):
+    """
+    Check if a user has a permission in a course.
+    """
+    if not use_permission_checks():
+        return False
+    elif isinstance(user, AnonymousUser):
+        return False
+    if isinstance(permission, CourseRolesPermission):
+        permission = permission.value.name
+    return UserRole.objects.filter(
+        user=user,
+        role__permissions__name=permission,
+        course=course_key,
+    ).exists()
+
+
+@request_cached()
+def user_has_permission_list_course(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        course_key: CourseKey
+):
+    """
+    Check if a user has all of the given permissions in a course.
+    """
+    if not use_permission_checks():
+        return False
+    return all(user_has_permission_course(user, permission, course_key) for permission in permissions)
+
+
+@request_cached()
+def user_has_permission_list_any_course(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        course_key: CourseKey
+):
+    """
+    Check if a user has ANY of the given permissions in a course.
+    """
+    return any(user_has_permission_course(user, permission, course_key) for permission in permissions)
+
+
+@request_cached()
+def user_has_permission_org(
+        user: Union[User, AnonymousUser],
+        permission: Union[CourseRolesPermission, str],
+        organization_name: str
+):
+    """
+    Check if a user has a permission for all courses in an organization.
+    """
+    if not use_permission_checks():
+        return False
+    elif isinstance(user, AnonymousUser):
+        return False
+    if isinstance(permission, CourseRolesPermission):
+        permission = permission.value.name
+    return UserRole.objects.filter(
+        user=user,
+        role__permissions__name=permission,
+        course__isnull=True,
+        org__name=organization_name,
+    ).exists()
+
+
+@request_cached()
+def user_has_permission_list_org(
+        user: Union[User, AnonymousUser],
+        permissions: List[Union[CourseRolesPermission, str]],
+        organization_name: str
```

Comment:
Given that permissions may become a long list of things, I think it's more intuitive for the argument order for these functions to be `user, (org | course), (permission | permissions)`.